### PR TITLE
Add rest api endpoints

### PR DESCRIPTION
**Pull request by joshbetz**
See #50

### CODE REVIEW

**Comment by ethitter** [comment]
In addition to the `wpcom_vip_go_rest_api_request_allowed()` call, we'll need some protection for the REST endpoints being blocked via the `rest_authentication_errors` filter. Cron Control handles it here: https://github.com/Automattic/vip-go-mu-plugins/blob/6d9c22a536fc62537dea797f621cd17888ffae10/001-cron.php#L45. We do the same for the sites endpoint here: https://github.com/Automattic/vip-go-mu-plugins/blob/d34e2e93e11135dbccbac573add05e354b988276/rest-api/vip-endpoints.php#L63.

**Comment by mjangda** [comment]
> we'll need some protection for the REST endpoints being blocked via the rest_authentication_errors filter.

Can we do this in a somewhat generic way so we don't have to repeat the same code everywhere?

**Comment by ethitter** [comment]
> Can we do this in a somewhat generic way so we don't have to repeat the same code everywhere?

Good call. The protection in https://github.com/Automattic/vip-go-mu-plugins/blob/d34e2e93e11135dbccbac573add05e354b988276/rest-api/vip-endpoints.php#L63 can be generalized to cover the entire `vip` namespace.

File: class-vip-support-cli.php [hunk 1]
```diff
@@ -115,28 +66,19 @@ public function remove_support_user( $args, $assoc_args ) {
 
 		$user_email = $args[0];
 
-		// Let's find the user
-		$user = get_user_by( 'email', $user_email );
+		$success = remove_support_user( $user_email );
```
Comment (mjangda) [inline class-vip-support-cli.php]: Same as `new_support_user`; we need a better name or use actual namespaces.

File: class-vip-support-cli.php [hunk 2]
```diff
@@ -115,28 +66,19 @@ public function remove_support_user( $args, $assoc_args ) {
...
 
-		if ( false === $user ) {
-			\WP_CLI::warning( "No user exists with the email address {$user_email}, so they could not be deleted" );
-			return;
+		if ( is_wp_error( $success ) && $success['code'] === 'invalid-user' ) {
+			return \WP_CLI::warning( "No user exists with the email address {$user_email}, so they could not be deleted" );
```
Comment (mjangda) [inline class-vip-support-cli.php]: Can `remove_support_user` just return this error message? Would avoid the need for the conditionals to check to the code.

File: class-vip-support-user.php [hunk 3]
```diff
@@ -788,3 +788,86 @@ protected function demote_user_from_vip_support_to( $user_id, $revert_role_to )
...
+
+	// If this is a multisite, commence super powers!
+	if ( is_multisite() ) {
+		grant_super_admin( $user->ID );
+	}
+
+	return $user_id;
+}
+
+function remove_support_user( $user_id, $by = 'email' ) {
```
Comment (mjangda) [inline class-vip-support-user.php]: Why not just make this and `create` static methods inside `WPCOM_VIP_Support_User`?

File: class-vip-support-api.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+	private static $namespace;
+
+	function __construct() {
+		self::$namespace = 'vip/v1';
+		add_action( 'rest_api_init', [ $this, 'rest_api_init' ] );
+	}
+
+	function permission_callback() {
+		return wpcom_vip_go_rest_api_request_allowed( self::$namespace );
```
Comment (mjangda) [inline class-vip-support-api.php]: Probably becomes moot if we move this library into mu-plugins but should we use a filter here? Or maybe add a `function_exists` check?

Comment (mjangda) [inline class-vip-support-api.php]: Do we need other protections like proxy / a8c IP only?

File: class-vip-support-api.php [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+				'methods' => WP_REST_Server::CREATABLE,
+				'permission_callback' => array( $this, 'permission_callback' ),
+				'args' => array(
+					'user_login' => array(
+						'required' => true,
+					),
+					'user_pass' => array(
+						'required' => true,
+						'validate_callback' => function( $param, $request, $key ) {
+							if ( strlen( $param ) < 10 ) {
```
Comment (mjangda) [inline class-vip-support-api.php]: `-data 'user_pass=aaaaaaaaaa'` <= super secure ;)

(No change needed here; this is fine)

**Comment by ethitter** [comment]
> Good call. The protection in https://github.com/Automattic/vip-go-mu-plugins/blob/d34e2e93e11135dbccbac573add05e354b988276/rest-api/vip-endpoints.php#L63 can be generalized to cover the entire `vip` namespace.

Done in https://github.com/Automattic/vip-go-mu-plugins/pull/749, along with adding support for query-string permalinks (because I was there ¯\_(ツ)_/¯).

File: class-vip-support-api.php (hunk 4 above)
Comment (ethitter) [reply class-vip-support-api.php]: We don't have a way to check if it's an a8c IP, only if the request is proxied. I'm not sure if that will be too restrictive, as a lot of internal tools will make requests from a8c IPs, but not via proxies (async jobs or the Go API, for example).

File: class-vip-support-cli.php [hunk 6]
```diff
@@ -1,13 +1,18 @@
...
 
 /**
  * Implements a WP CLI command that converts guid users to meta users
  * Class command
  *
  * @package a8c\vip_support
  */
-class WPCOM_VIP_Support_CLI  extends WP_CLI_Command {
-
+class CLI extends WP_CLI_Command {
```
Comment (mjangda) [inline class-vip-support-cli.php]: Let's rename this to `Command` or `CLI_Command`

File: class-vip-support-cli.php [hunk 7]
```diff
@@ -40,62 +45,13 @@ public function add_support_user( $args, $assoc_args ) {
...
-		if ( is_multisite() ) {
-			grant_super_admin( $user->ID );
-		}
-
-		$msg = "Added";
-		if ( $update_user ) {
-			$msg = "Updated";
-		}
-		$msg .= " user $user_id with login {$user_login} and password '{$user_pass}', they are verified as a VIP Support user and ready to go";
+		$msg = "Added user $user_id with login {$user_login} and password '{$user_pass}', they are verified as a VIP Support user and ready to go";
```
Comment (mjangda) [inline class-vip-support-cli.php]: I know this is old code, but we should stop outputting the password :)

File: class-vip-support-user.php [hunk 8]
```diff
@@ -654,7 +661,7 @@ public static function is_verified_automattician( $user_id ) {
 		return ( $is_a8c_email && $email_verified );
 	}
 
-	public function user_has_vip_support_role( $user_id, $active_role = true ) {
+	public static function user_has_vip_support_role( $user_id, $active_role = true ) {
```
Comment (mjangda) [inline class-vip-support-user.php]: Not a big deal but feels somewhat inconsistent to make this static but still have a method like `user_has_verified_email` which is not static.

**Comment by mjangda** [comment]
Oh, I guess tests are failing here too. We should definitely fix before merging.

**Comment by ethitter** [comment]
Yeah, the tests have been broken for...a while. Travis wasn't even enabled until yesterday. :)

File: class-vip-support-user.php (hunk 8 above)
Comment (ethitter) [reply class-vip-support-user.php]: Yeah, it's a bit strange, but the method doesn't do anything with the instance, and this was needed to access it from the new `remove()` method. A quick glance indicates that most of the public methods could be made static.

**Comment by ethitter** [comment]
Tests are fixed in #54, which is meant to be merged into this.

**Pull request merged**